monkey/common/di_container.py

```diff
@@ -0,0 +1,122 @@
+import inspect
+from typing import Any, MutableMapping, Sequence, Type, TypeVar
+
+T = TypeVar("T")
+
+
+class UnregisteredTypeError(ValueError):
+    pass
+
+
+class DIContainer:
+    """
+    A dependency injection (DI) container that uses type annotations to resolve and inject
+    dependencies.
+    """
+
+    def __init__(self):
+        self._type_registry = {}
+        self._instance_registry = {}
+
+    def register(self, interface: Type[T], concrete_type: Type[T]):
```

Comment:
This should be called `register_class`?

Reply:
So all services need to be classes?

Reply:
Yes, but they already are, though some of them pretend they're not. Others are really just utility functions and don't manipulate the data model in any way, so they're not really services and should be moved into a different package/module that contains utility functions.

Reply:
> This should be called `register_class`?

[Register, Resolve, Release (RRR)](https://blog.ploeh.dk/2010/09/29/TheRegisterResolveReleasepattern/) is an established pattern. I'd prefer to just leave it as `register()` for that reason.